compiler/src/iree/compiler/Modules/IO/Parameters/Transforms/ParameterizeGlobals.cpp

```diff
@@ -0,0 +1,248 @@
+    }
+
+    // Allocate an index we'll populate during building to allow us to get the
+    // storage ranges of non-metadata parameters.
+    iree_io_parameter_index_t *built_index = NULL;
+    status = iree_io_parameter_index_create(host_allocator, &built_index);
+    auto releaseIndexExit = llvm::make_scope_exit(
+        [&]() { return iree_io_parameter_index_release(built_index); });
+    if (!iree_status_is_ok(status)) {
+      moduleOp.emitError() << "Failed to allocate parameter index";
+      return signalPassFailure();
+    }
+
+    // Commit the archive header to the file and produce an index referencing
+    // it. This will allow us to know where to copy file contents.
+    status = iree_io_parameter_archive_builder_write(
+        &builder, target_file_handle, target_file_offset, target_stream,
+        built_index);
+    if (!iree_status_is_ok(status)) {
+      moduleOp.emitError() << "Failed to write parameter index header to "
+                           << "output file " << archivePath;
+      return signalPassFailure();
+    }
+
+    StringAttr namespaceAttr =
+        parameterNamespace.empty()
+            ? StringAttr()
+            : StringAttr::get(context, parameterNamespace);
+
+    // Write all of the global contents to the appropriate data storage
+    // segments.
+    for (auto constantGlobal : constantGlobals) {
+      StringRef name = constantGlobal.getSymName();
+
+      const iree_io_parameter_index_entry_t *target_entry = NULL;
+      status = iree_io_parameter_index_lookup(
+          built_index,
+          iree_string_view_t{name.data(),
+                             static_cast<iree_host_size_t>(name.size())},
+          &target_entry);
+      if (!iree_status_is_ok(status)) {
+        moduleOp.emitError() << "Failed to write parameter index header to "
+                             << "output file " << archivePath;
+        return signalPassFailure();
+      }
+      status = iree_io_stream_seek(target_stream, IREE_IO_STREAM_SEEK_SET,
+                                   target_file_offset +
+                                       target_entry->storage.file.offset);
+      if (!iree_status_is_ok(status)) {
+        moduleOp.emitError() << "Failed to seek to location of global in index "
+                             << constantGlobal;
+        return signalPassFailure();
+      }
+
+      auto initialValueAttr = constantGlobal.getInitialValueAttr();
+      auto serializableAttr =
+          dyn_cast<IREE::Util::SerializableAttrInterface>(initialValueAttr);
+
+      SmallVector<char> buffer;
+      if (failed(serializableAttr.serializeToVector(
+              constantGlobal.getLoc(), llvm::endianness::native, buffer))) {
+        moduleOp.emitError() << "Failed to serialize global " << constantGlobal;
+        return signalPassFailure();
+      }
+
+      status =
+          iree_io_stream_write(target_stream, buffer.size(),
+                               reinterpret_cast<const char *>(buffer.data()));
```

Comment:
to avoid the allocation and copy here (which we really want to do with large parameters) we can have an llvm::raw_ostream implementation that wraps iree_io_stream_t - then we can use `serializableAttr.serializeToStream` to zero-copy write the results.

something like:
```c++
class iree_io_stream_ostream : public llvm::raw_ostream {
 public:
  explicit iree_io_stream_ostream(iree_io_stream_t* stream) : stream(stream) {
    iree_io_stream_retain(stream);
  }
  ~iree_io_stream_ostream() override {
    iree_io_stream_release(stream);
  }
 private:
  uint64_t current_pos() const override { return iree_io_stream_offset(stream); }
  void write_impl(const char *Ptr, size_t Size) override {
    IREE_CHECK_OK(iree_io_stream_write(stream, ptr, size));
  }
  iree_io_stream_t* stream = NULL;
};
```